test/saithrift/tests/saiacl.py

```diff
@@ -15,114 +15,427 @@
+###############################################################################################################
 
 @group('acl')
-class IPAclTest(sai_base_test.ThriftInterfaceDataPlane):
-    def runTest(self):
-        print
-        print "Sending packet port 1 -> port 2 (192.168.0.1 -> 10.10.10.1 [id = 101])"
-        switch_init(self.client)
-        port1 = port_list[1]
-        port2 = port_list[2]
+class aclTestBase(object):
+
+    def setupRouting(self, port1, port2, port3):
         v4_enabled = 1
         v6_enabled = 1
         mac = ''
 
-        vr_id = sai_thrift_create_virtual_router(self.client, v4_enabled, v6_enabled)
-
-        rif_id1 = sai_thrift_create_router_interface(self.client, vr_id, 1, port1, 0, v4_enabled, v6_enabled, mac)
-        rif_id2 = sai_thrift_create_router_interface(self.client, vr_id, 1, port2, 0, v4_enabled, v6_enabled, mac)
-
-        addr_family = SAI_IP_ADDR_FAMILY_IPV4
-        ip_addr1 = '10.10.10.1'
-        ip_mask1 = '255.255.255.255'
-        dmac1 = '00:11:22:33:44:55'
-        sai_thrift_create_neighbor(self.client, addr_family, rif_id1, ip_addr1, dmac1)
-        nhop1 = sai_thrift_create_nhop(self.client, addr_family, ip_addr1, rif_id1)
-        sai_thrift_create_route(self.client, vr_id, addr_family, ip_addr1, ip_mask1, nhop1)
-
-        # send the test packet(s)
-        pkt = simple_tcp_packet(eth_dst='00:77:66:55:44:33',
-                                eth_src='00:22:22:22:22:22',
-                                ip_dst='10.10.10.1',
-                                ip_src='192.168.0.1',
-                                ip_id=105,
-                                ip_ttl=64)
-        exp_pkt = simple_tcp_packet(
-                                eth_dst='00:11:22:33:44:55',
-                                eth_src='00:77:66:55:44:33',
-                                ip_dst='10.10.10.1',
-                                ip_src='192.168.0.1',
-                                ip_id=105,
-                                ip_ttl=63)
+        self.vr_id = sai_thrift_create_virtual_router(self.client, v4_enabled, v6_enabled)
+
+        self.rif_id1 = sai_thrift_create_router_interface(self.client, self.vr_id, 1, port1, 0, v4_enabled, v6_enabled, mac)
+        self.rif_id2 = sai_thrift_create_router_interface(self.client, self.vr_id, 1, port2, 0, v4_enabled, v6_enabled, mac)
+        self.rif_id3 = sai_thrift_create_router_interface(self.client, self.vr_id, 1, port3, 0, v4_enabled, v6_enabled, mac)
+
+        self.addr_family = SAI_IP_ADDR_FAMILY_IPV4
+        self.ip_addr1 = IP_DST
+        self.ip_mask1 = IP_MASK
+        self.dmac1 = rewrite_mac2
+        sai_thrift_create_neighbor(self.client, self.addr_family, self.rif_id1, self.ip_addr1, self.dmac1)
+        self.nhop1 = sai_thrift_create_nhop(self.client, self.addr_family, self.ip_addr1, self.rif_id1)
+        sai_thrift_create_route(self.client, self.vr_id, self.addr_family, self.ip_addr1, self.ip_mask1, self.nhop1)
```

Comment:
no, you do not need. In the sai pipeline, there is l3 routing table, than after that there is neighbor table. In line 69, you created neighbor entry in the neighbor table. Only thing you need to add is in the routing table, you add a directly reachable route. You do not line 70.

Routing table: 10.10.10.0/255.255.255.0 -> rif_id1
Neighbor table: (rif_id1, 10.10.10.1) - > dmac

When a packet with dip=10.10.10.1 comes in, it first match the routing table, and then match the neighbor table to find out the dmac.

---
In reply to: [78834714](https://github.com/opencomputeproject/SAI/pull/148#discussion_r78834714) [](ancestors = 78834714)